File: resultcapture-service/src/main/java/com/devotion/capture/SurveyCaptureApplication.java
package com.devotion.capture;

import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.*;
import org.springframework.kafka.support.converter.StringJsonMessageConverter;
import springfox.documentation.builders.ApiInfoBuilder;
import springfox.documentation.builders.RequestHandlerSelectors;
import springfox.documentation.service.ApiInfo;
import springfox.documentation.service.Contact;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.spring.web.plugins.Docket;
import springfox.documentation.swagger2.annotations.EnableSwagger2;

import java.util.HashMap;
import java.util.Map;


@SpringBootApplication
@EnableSwagger2
public class SurveyCaptureApplication {

    @Value("${kafka.bootstrap-servers}")
    private String bootstrapAddress;

    @Value("${kafka.consumer-group}")
    private String consumerGroupName;


    @Value("${api.version}")
    private String apiVersion;

    public static void main(String[] args) {
        SpringApplication.run(SurveyCaptureApplication.class, args);
    }

    @Bean
    public ModelMapper modelMapper() {
        return new ModelMapper();
    }

    @Bean
    public Docket api() {
        //noinspection Guava
        return new Docket(DocumentationType.SWAGGER_2)
                .select()
                .apis(RequestHandlerSelectors.basePackage("com.devotion"))
                .build().apiInfo(metadata());
    }

    @Bean
    public KafkaTemplate<String, String> kafkaTemplate(ProducerFactory<String, String> producerFactory) {
        KafkaTemplate<String, String> kafkaTemplate = new KafkaTemplate<>(producerFactory);
        kafkaTemplate.setMessageConverter(new StringJsonMessageConverter());
        return kafkaTemplate;
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, String> jsonKafkaListenerContainerFactory() {
        ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(consumerFactory());
        factory.setMessageConverter(new StringJsonMessageConverter());
        return factory;
    }

    @Bean
    public ConsumerFactory<String, String> consumerFactory() {
        return new DefaultKafkaConsumerFactory<>(consumerProperties());
    }

    @Bean
    public Map<String, Object> consumerProperties() {
        Map<String, Object> props = new HashMap<>();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapAddress);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupName);
        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
        props.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, 15000);
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        return props;
    }

    @Bean
    public ProducerFactory<String, String> producerFactory() {
        Map<String, Object> props = new HashMap<>();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapAddress);
        props.put(ProducerConfig.RETRIES_CONFIG, 0);
        props.put(ProducerConfig.BATCH_SIZE_CONFIG, 16384);
        props.put(ProducerConfig.LINGER_MS_CONFIG, 1);
        props.put(ProducerConfig.BUFFER_MEMORY_CONFIG, 33554432);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        return new DefaultKafkaProducerFactory<>(props);
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory() {
        ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(consumerFactory());
        return factory;
    }

    /*  @Bean
      public ProducerFactory<String, Message> producerFactory() {
          Map<String, Object> configProps = getConfigProps();
          return new DefaultKafkaProducerFactory<>(configProps);
      }


      @Bean
      public KafkaTemplate<String, CaptureResultCreatedEvent> kafkaTemplate() {
          return new KafkaTemplate<>(producerFactory());
      }


      @Bean
      public ConcurrentKafkaListenerContainerFactory<String, CaptureResultCreatedEvent> kafkaListenerContainerFactory() {
          ConcurrentKafkaListenerContainerFactory<String, CaptureResultCreatedEvent> factory
                  = new ConcurrentKafkaListenerContainerFactory<>();
          factory.setConsumerFactory(consumerFactory());

          return factory;
      }

      @Bean
      public ConsumerFactory<String, CaptureResultCreatedEvent> consumerFactory() {
          JsonDeserializer<CaptureResultCreatedEvent> deserializer = new JsonDeserializer<>(CaptureResultCreatedEvent.class);
          Map<String, String> conf = new HashMap<>();
          conf.put(JsonDeserializer.TRUSTED_PACKAGES,"*");
          deserializer.configure(conf, false);
          return new DefaultKafkaConsumerFactory<>(
                  getConfigProps(),
                  new StringDeserializer(),
                  deserializer);
      }

      private Map<String, Object> getConfigProps() {
          Map<String, Object> configProps = new HashMap<>();
          configProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapAddress);
          configProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
          configProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class);

          configProps.put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupName);
          configProps.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, true);
          configProps.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, "100");
          configProps.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "15000");
          configProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringSerializer.class);
          configProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonSerializer.class);

          return configProps;
      }
  */
    private ApiInfo metadata() {
        return new ApiInfoBuilder()
                .title("Survey submission API")
                .description("Operations on this API allows you to capture results of survey submission.")
                .contact(new Contact("Dragan Ljubojevic", "", "dragan.ljubojevic@gmail.com"))
                .version(apiVersion)
                .build();
    }

}

File: resultcapture-service/src/main/kotlin/com/devotion/capture/SurveyCaptureApplication.kt
package com.devotion.capture

import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.kafka.common.serialization.StringSerializer
import org.modelmapper.ModelMapper
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.annotation.Bean
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.*
import org.springframework.kafka.support.converter.StringJsonMessageConverter
import springfox.documentation.builders.ApiInfoBuilder
import springfox.documentation.builders.RequestHandlerSelectors
import springfox.documentation.service.Contact
import springfox.documentation.spi.DocumentationType
import springfox.documentation.spring.web.plugins.Docket
import springfox.documentation.swagger2.annotations.EnableSwagger2
import java.util.*


@SpringBootApplication
@EnableSwagger2
open class SurveyCaptureApplication {

    @Value("\${kafka.bootstrap-servers}")
    private lateinit var bootstrapAddress: String

    @Value("\${kafka.consumer-group}")
    private lateinit var consumerGroupName: String

    @Value("\${api.version}")
    private lateinit var apiVersion: String

    @Bean
    open fun modelMapper() = ModelMapper()

    @Bean
    open fun api(): Docket = Docket(DocumentationType.SWAGGER_2)
            .select()
            .apis(RequestHandlerSelectors.basePackage("com.devotion"))
            .build()
            .apiInfo(apiInfo())

    @Bean
    open fun kafkaTemplate(producerFactory: ProducerFactory<String, String>): KafkaTemplate<String, String> {
        val kafkaTemplate = KafkaTemplate(producerFactory)
        kafkaTemplate.setMessageConverter(StringJsonMessageConverter())
        return kafkaTemplate
    }

    @Bean
    open fun jsonKafkaListenerContainerFactory(): ConcurrentKafkaListenerContainerFactory<String, String> {
        val factory = ConcurrentKafkaListenerContainerFactory<String, String>()
        factory.consumerFactory = consumerFactory()
        factory.setMessageConverter(StringJsonMessageConverter())
        return factory
    }

    @Bean
    open fun consumerFactory() = DefaultKafkaConsumerFactory<String, String>(consumerProperties())

    @Bean
    open fun consumerProperties() = HashMap<String, Any>().apply {
        put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapAddress)
        put(ConsumerConfig.GROUP_ID_CONFIG, consumerGroupName)
        put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
        put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, 15000)
        put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer::class.java)
        put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer::class.java)
    }

    @Bean
    open fun producerFactory() = DefaultKafkaProducerFactory<String, String>(
            HashMap<String, Any>().apply {
                put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapAddress)
                put(ProducerConfig.RETRIES_CONFIG, 0)
                put(ProducerConfig.BATCH_SIZE_CONFIG, 16384)
                put(ProducerConfig.LINGER_MS_CONFIG, 1)
                put(ProducerConfig.BUFFER_MEMORY_CONFIG, 33554432)
                put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer::class.java)
                put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer::class.java)
            }
    )

    @Bean
    open fun kafkaListenerContainerFactory(): ConcurrentKafkaListenerContainerFactory<String, String> {
        val factory = ConcurrentKafkaListenerContainerFactory<String, String>()
        factory.consumerFactory = consumerFactory()
        return factory
    }

    private fun apiInfo() = ApiInfoBuilder()
            .title("Survey submission API")
            .description("Operations on this API allows you to capture results of survey submission.")
            .contact(Contact("Dragan Ljubojevic", "", "dragan.ljubojevic@gmail.com"))
            .version(apiVersion)
            .build()

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            SpringApplication.run(SurveyCaptureApplication::class.java, *args)
        }
    }
}
